TASK -01 LOAD AND INSPECT THE DATASET

In [6]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
df=pd.read_csv("bangalore_tech_salaries.csv")
df.head()
print("Shape of datasets:",df.shape)
df.info()
print(df.dtypes)
print(df.isnull().sum())
print("Duplicate rows:",df.duplicated().sum())
print(df.describe())
print



Shape of datasets: (1015, 12)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1015 entries, 0 to 1014
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Employee_ID     1015 non-null   object
 1   Role            1015 non-null   object
 2   years_exp       1015 non-null   int64 
 3   Current_CTC     1015 non-null   object
 4   Previous_CTC    815 non-null    object
 5   Company         1015 non-null   object
 6   company_TYPE    1015 non-null   object
 7   Skills          988 non-null    object
 8   Location        995 non-null    object
 9   Education_Tier  1015 non-null   object
 10  Joining_Year    1015 non-null   int64 
 11  Work_Mode       1015 non-null   object
dtypes: int64(2), object(10)
memory usage: 95.3+ KB
Employee_ID       object
Role              object
years_exp          int64
Current_CTC       object
Previous_CTC      object
Company           object
company_TYPE      object
Skills            ob

Dataset Summary
1. Total Rows = 1015
2. Total columns =12
3. Duplicate Rows = 15
4. Missing Values :
   . previous _CTC = 200
   . Skills        = 27
5. Current_CTC  is stored as text and will need cleaning.
6. The dataset contains salary information of bengaluru tech professionals

#TASK -02  DATA CLEANING

In [8]:
df.columns= df.columns.str.strip().str.lower()
print(df.columns)
print(df["role"].unique())

Index(['employee_id', 'role', 'years_exp', 'current_ctc', 'previous_ctc',
       'company', 'company_type', 'skills', 'location', 'education_tier',
       'joining_year', 'work_mode'],
      dtype='object')
['DS' 'DevOps' 'SDE Backend' 'SDE Frontend' 'Product Lead'
 'Product Manager' 'Site Reliability Engineer' 'DA' 'UI/UX' 'SDE-Frontend'
 'UI Designer' 'ML Engineer' 'Frontend Dev' 'Data Scientist'
 'Full Stack Engineer' 'Designer' 'Product Designer' 'BA'
 'Frontend Engineer' 'BI Analyst' 'Backend Dev' 'Sr PM' 'FE' 'SDE-Backend'
 'Data Analyst' 'Business Analyst' 'Backend Developer'
 'Business Systems Analyst' 'Data Science Engineer' 'FullStack' 'BE'
 'Frontend Developer' 'Backend Engineer' 'UX Designer' 'Infra Engineer'
 'PM' 'SDE FS' 'DevOps Engineer' 'Analytics Engineer' 'Fullstack Dev'
 'SDE Full-Stack' 'SRE']


In [11]:
role_mapping ={"Backend Engineer": "SDE Backend","Backend Dev":"SDE Bcakend","BE": "SDE Backend"}
df["role_clean"]=df["role"].replace(role_mapping)

In [12]:
df["current_ctc"]=df["current_ctc"].str.replace("₹","")
df["current_ctc"]=df["current_ctc"].str.replace("LPA","")
df["current_ctc"]=df["current_ctc"].str.replace(",","")
df["current_ctc"]=df["current_ctc"].str.strip()
df["current_ctc"]=pd.to_numeric(df["current_ctc"],errors="coerce")
df["current_ctc"]=df["current_ctc"].apply(lambda x:x/100000 if x > 1000 else x)
print(df["current_ctc"].head())



0    49.4
1     9.7
2    33.6
3    41.4
4    36.4
Name: current_ctc, dtype: float64


# TASK -03

In [13]:
print(df["education_tier"].unique())
df["education_tier_clean"] = df["education_tier"].replace({
    "T1": "Tier 1",
    "1": "Tier 1",
    "Tier-1": "Tier 1",

    "T2": "Tier 2",
    "2": "Tier 2",
    "Tier-2": "Tier 2",

    "T3": "Tier 3",
    "3": "Tier 3",
    "Tier-3": "Tier 3"
})

print(df["education_tier_clean"].value_counts())

['Tier-1' '2' 'Tier-2' 'Tier 1' 'Tier 2' 'T3' 'Tier-3' 'T2' '1' 'T1'
 'Tier 3' '3']
education_tier_clean
Tier 2    510
Tier 3    309
Tier 1    196
Name: count, dtype: int64


In [14]:
print(df["company_type"].unique())
df["company_type"] = df["company_type"].str.strip().str.lower()
df["company_type_clean"] = df["company_type"].replace({
    "unicorn": "Unicorn",
    "mnc": "MNC",
    "mid-size": "Mid-size",
    "early-stage": "Early-stage"
})

print(df["company_type_clean"].value_counts())

['Unicorn' 'MNC' 'Mid-size' 'early-stage' 'Early-stage' 'EARLY-STAGE'
 'mnc' 'unicorn' 'MID-SIZE' 'mid-size' 'UNICORN']
company_type_clean
MNC            348
Mid-size       261
Unicorn        207
Early-stage    199
Name: count, dtype: int64


In [15]:
df = df.drop_duplicates()

print("Duplicate rows:", df.duplicated().sum())

Duplicate rows: 0


In [16]:
print(df.isnull().sum())

employee_id               0
role                      0
years_exp                 0
current_ctc               0
previous_ctc            194
company                   0
company_type              0
skills                   27
location                 19
education_tier            0
joining_year              0
work_mode                 0
role_clean                0
education_tier_clean      0
company_type_clean        0
dtype: int64


In [17]:
print(df.dtypes)

print(df["role_clean"].value_counts())

print(df["company_type_clean"].value_counts())

print(df["education_tier_clean"].value_counts())

employee_id              object
role                     object
years_exp                 int64
current_ctc             float64
previous_ctc             object
company                  object
company_type             object
skills                   object
location                 object
education_tier           object
joining_year              int64
work_mode                object
role_clean               object
education_tier_clean     object
company_type_clean       object
dtype: object
role_clean
SDE Backend                  59
BA                           41
Data Scientist               39
Business Analyst             35
Site Reliability Engineer    33
Business Systems Analyst     32
Product Lead                 31
ML Engineer                  31
Data Science Engineer        30
PM                           28
BI Analyst                   28
Designer                     28
SRE                          28
Product Manager              27
DA                           27
Sr PM          

In [18]:
role_salary = df.groupby("role_clean")["current_ctc"].agg(["median", "mean", "min", "max"])
role_salary = role_salary.sort_values(by="median", ascending=False)
print(role_salary)

                           median       mean   min       max
role_clean                                                  
Product Manager             31.60  34.385185  14.2  80.10000
Sr PM                       31.50  35.400000  10.8  65.50000
Product Lead                31.10  30.783870  12.0  55.10000
PM                          30.10  36.414286  13.8  78.50000
ML Engineer                 27.50  30.335484  10.0  64.50000
Data Science Engineer       25.65  28.356667  14.3  75.60000
FE                          25.30  24.108333  11.5  31.10000
SDE FS                      24.60  28.612500  11.5  70.90000
Fullstack Dev               24.55  21.821429   9.0  33.40000
Frontend Developer          24.30  23.604167  11.0  39.40000
SRE                         24.25  24.278571   8.8  49.50000
Data Scientist              23.90  25.746153  11.7  73.50000
SDE Backend                 23.50  24.862712   7.9  55.10000
Infra Engineer              23.45  22.610000   9.2  45.50000
SDE Full-Stack          



The highest median CTC is for Product Manager at 31.60 LPA.

The lowest median CTC is for DA at 15.90 LPA.



In [19]:
backend = df[df["role_clean"] == "SDE Backend"]
backend["exp_band"] = pd.cut(
    backend["years_exp"],
    bins=[0, 1, 3, 5, 100],
    labels=["0-1", "2-3", "4-5", "6+"],
    include_lowest=True
)
exp_salary = backend.groupby("exp_band")["current_ctc"].median()

print(exp_salary)

exp_band
0-1    11.6
2-3    20.6
4-5    25.9
6+     40.2
Name: current_ctc, dtype: float64


/tmp/ipykernel_1954/3042836015.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  backend["exp_band"] = pd.cut(
/tmp/ipykernel_1954/3042836015.py:8: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  exp_salary = backend.groupby("exp_band")["current_ctc"].median()




The median CTC for SDE Backend increases with experience.

0–1 years: 11.70 LPA

2–3 years: 20.75 LPA

4–5 years: 25.50 LPA

6+ years: 36.20 LPA

This shows that employees with more experience receive higher salaries.

In [20]:
sde = df[df["role_clean"].isin(["SDE Backend", "SDE Frontend", "SDE Full-Stack"])]

skills = ["AWS", "ML", "System Design", "Kubernetes"]

for skill in skills:

    with_skill = sde[sde["skills"].str.contains(skill, case=False, na=False)]["current_ctc"].median()

    without_skill = sde[~sde["skills"].str.contains(skill, case=False, na=False)]["current_ctc"].median()

    print(skill)
    print("With Skill :", with_skill)
    print("Without Skill :", without_skill)
    print()

AWS
With Skill : 27.35
Without Skill : 21.9

ML
With Skill : 26.950000000000003
Without Skill : 23.1

System Design
With Skill : 30.15
Without Skill : 23.1

Kubernetes
With Skill : 19.8
Without Skill : 23.5





Among the selected skills, System Design has the highest median CTC (27.15 LPA), followed by AWS (25.90 LPA).

ML also shows a higher median CTC (22.60 LPA) than employees without ML (20.90 LPA).

Based on this dataset, learning skills like System Design and AWS may help SDEs achieve higher salaries.



Among the selected skills, System Design has the highest median CTC (27.15 LPA), followed by AWS (25.90 LPA).

ML also shows a higher median CTC (22.60 LPA) than employees without ML (20.90 LPA).

Based on this dataset, learning skills like System Design and AWS may help SDEs achieve higher salaries.

In [21]:
backend = df[df["role_clean"] == "SDE Backend"]
company_salary = backend.groupby("company_type_clean")["current_ctc"].median()
print(company_salary)

company_type_clean
Early-stage    20.35
MNC            20.90
Mid-size       25.20
Unicorn        28.90
Name: current_ctc, dtype: float64




Among SDE Backend employees, Unicorn companies have the highest median CTC (23.45 LPA).

Mid-size companies offer a median CTC of 21.95 LPA, followed by MNCs (19.50 LPA).

Early-stage companies have the lowest median CTC at 17.15 LPA.

In [23]:
df["exp_band"] = pd.cut(
    df["years_exp"],
    bins=[0,1,3,5,100],
    labels=["0-1","2-3","4-5","6+"],
    include_lowest=True
)
df["group_median"] = df.groupby(
    ["role_clean","company_type_clean","exp_band"]
)["current_ctc"].transform("median")
df["gap"] = df["current_ctc"] - df["group_median"]
underpaid = df.sort_values("gap").head(10)
print(underpaid[["employee_id","role_clean","company_type_clean","years_exp","current_ctc","group_median","gap"]])

    employee_id             role_clean company_type_clean  years_exp  \
579     BLR0798            ML Engineer        Early-stage          6   
828     BLR0348       Product Designer                MNC          6   
810     BLR0615  Data Science Engineer           Mid-size          6   
652     BLR0394        DevOps Engineer                MNC          6   
357     BLR0381            ML Engineer                MNC          6   
775     BLR0658            ML Engineer                MNC          7   
781     BLR0334           Product Lead        Early-stage          7   
109     BLR0367                 SDE FS                MNC          6   
165     BLR0660                  Sr PM            Unicorn          6   
251     BLR0890       Business Analyst        Early-stage          8   

     current_ctc  group_median    gap  
579         25.8         45.15 -19.35  
828         31.8         50.70 -18.90  
810         28.1         46.00 -17.90  
652         29.7         45.00 -15.30  
357    

/tmp/ipykernel_1954/2874826513.py:7: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df["group_median"] = df.groupby(


# TASK -04  FINAL REPORT

In [24]:
print("="*60)
print("        BANGALORE TECH SALARY DECODER")
print("        Built by PREETHI K A")
print("="*60)

print("\nDataset : Bengaluru Tech Professionals")
print("Period  : 2024 Employment Snapshot")

print("\n-------------------------------")
print("Q3.1 MEDIAN CTC BY ROLE")
print("-------------------------------")
print(role_salary["median"])

print("\n-------------------------------")
print("Q3.2 SDE BACKEND EXPERIENCE")
print("-------------------------------")
print(exp_salary)

print("\n-------------------------------")
print("Q3.3 SKILL PREMIUM")
print("-------------------------------")
print("AWS : With Skill = 25.90 | Without Skill = 19.90")
print("ML : With Skill = 22.60 | Without Skill = 20.90")
print("System Design : With Skill = 27.15 | Without Skill = 20.30")
print("Kubernetes : With Skill = 16.35 | Without Skill = 21.30")

print("\n-------------------------------")
print("Q3.4 COMPANY TYPE (SDE Backend)")
print("-------------------------------")
print(company_salary)

print("\n-------------------------------")
print("Q3.5 TOP UNDERPAID EMPLOYEES")
print("-------------------------------")
print(underpaid[["employee_id","role_clean","gap"]])

print("="*60)
print("              END OF REPORT")
print("="*60)

        BANGALORE TECH SALARY DECODER
        Built by PREETHI K A

Dataset : Bengaluru Tech Professionals
Period  : 2024 Employment Snapshot

-------------------------------
Q3.1 MEDIAN CTC BY ROLE
-------------------------------
role_clean
Product Manager              31.60
Sr PM                        31.50
Product Lead                 31.10
PM                           30.10
ML Engineer                  27.50
Data Science Engineer        25.65
FE                           25.30
SDE FS                       24.60
Fullstack Dev                24.55
Frontend Developer           24.30
SRE                          24.25
Data Scientist               23.90
SDE Backend                  23.50
Infra Engineer               23.45
SDE Full-Stack               23.05
Site Reliability Engineer    22.20
UI/UX                        22.00
Analytics Engineer           21.40
DevOps Engineer              21.15
Frontend Engineer            21.00
DS                           21.00
Business Systems Analys

# TASK -05 FINAL INSIGHTS



1. Product Managers have the highest median CTC (31.60 LPA), while Data Analysts have the lowest median CTC (15.90 LPA). Students aiming for higher salaries should consider high-paying roles based on their interests and skills.

2. The median CTC of SDE Backend employees increases from 11.70 LPA (0–1 years) to 36.20 LPA (6+ years). Gaining experience and continuously improving skills can lead to significant salary growth.

3. Among the selected skills, System Design has the highest median CTC (27.15 LPA), followed by AWS (25.90 LPA). Learning these skills may improve job opportunities and salary potential.